BÀI 1 

In [1]:
import torch
a = torch.tensor( 
    [[1,2,3],
     [4,5,6]],
    dtype=torch.int32
)
print (a)

tensor([[1, 2, 3],
        [4, 5, 6]], dtype=torch.int32)


In [2]:
print("Kích thước:", a.shape)
print("Số chiều:", a.ndim)
print("Kiểu dữ liệu:", a.dtype)

Kích thước: torch.Size([2, 3])
Số chiều: 2
Kiểu dữ liệu: torch.int32


BÀI 2

In [4]:
import torch

a = torch.rand(3, 4)
b = torch.randn(3, 4)
c = torch.zeros(2, 3)
d = torch.ones(2, 3)
e = torch.eye(4)

print("rand:")
print(a)

print("\nrandn:")
print(b)

print("\nzeros:")
print(c)

print("\nones:")
print(d)

print("\neye:")
print(e)


rand:
tensor([[0.9089, 0.7650, 0.9464, 0.2407],
        [0.8664, 0.5249, 0.8445, 0.4047],
        [0.1338, 0.9674, 0.7236, 0.9079]])

randn:
tensor([[ 0.1123, -0.0501, -1.7983, -0.9123],
        [-1.2731,  0.2253, -0.3950,  0.7868],
        [-0.3867,  0.8518, -1.7959, -0.6036]])

zeros:
tensor([[0., 0., 0.],
        [0., 0., 0.]])

ones:
tensor([[1., 1., 1.],
        [1., 1., 1.]])

eye:
tensor([[1., 0., 0., 0.],
        [0., 1., 0., 0.],
        [0., 0., 1., 0.],
        [0., 0., 0., 1.]])


Nhận xét
torch.rand() tạo các giá trị ngẫu nhiên theo phân phối đều, trong khoảng [0, 1).
torch.randn() tạo các giá trị ngẫu nhiên theo phân phối chuẩn (Normal), với trung bình gần 0 và độ lệch chuẩn gần 1.
Vì vậy, hai hàm đều tạo Tensor ngẫu nhiên nhưng khác nhau về phân phối của các giá trị được sinh ra.

In [6]:
f = torch.full((5, 5), 10)

print(f)

tensor([[10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10]])


BÀI 3

In [7]:
import torch

a = torch.zeros(2, 3, 4)

print("Tensor:")
print(a)

print("Shape:", a.shape)
print("Size:", a.size())
print("Number of dimensions:", a.ndim)
print("Data type:", a.dtype)

Tensor:
tensor([[[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]],

        [[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]]])
Shape: torch.Size([2, 3, 4])
Size: torch.Size([2, 3, 4])
Number of dimensions: 3
Data type: torch.float32


NHẬN XÉT: 
torch.Size([2, 3, 4]) cho biết Tensor a có 3 chiều, với kích thước lần lượt là 2, 3 và 4. Điều này có nghĩa Tensor gồm 2 × 3 × 4 = 24 phần tử.

BÀI 4

In [8]:
import torch
a = torch.randn(3, 4, 5)
print("Original tensor:")
print(a)
print("\nTensor a[1]:")
print(a[1])
print("\nTensor a[1:, 2:4]:")
print(a[1:, 2:4])

Original tensor:
tensor([[[ 0.5604, -0.3466, -0.1228,  1.5262,  0.5968],
         [ 0.6207,  1.1148,  0.7993, -2.9484, -0.5147],
         [-0.7107,  0.0273,  0.2386, -0.4894, -1.2085],
         [-1.1184,  1.2875, -0.2203,  0.9853, -0.5922]],

        [[ 2.1952, -0.3436, -0.6537,  0.3552,  0.7421],
         [ 0.0818,  2.1903,  0.8122, -0.7558, -0.2232],
         [ 0.4526, -1.0238, -1.7574,  0.4045,  0.0904],
         [-0.1112,  0.4386,  0.5655,  1.2266,  0.7068]],

        [[ 0.1982, -1.5306,  0.8350, -1.4998, -1.0209],
         [ 0.2995,  0.8333, -1.0496,  1.3314, -0.8596],
         [-0.7949, -0.2985,  1.9741, -1.4476, -0.9064],
         [ 0.0389,  0.4490,  0.8730, -0.2195,  1.2015]]])

Tensor a[1]:
tensor([[ 2.1952, -0.3436, -0.6537,  0.3552,  0.7421],
        [ 0.0818,  2.1903,  0.8122, -0.7558, -0.2232],
        [ 0.4526, -1.0238, -1.7574,  0.4045,  0.0904],
        [-0.1112,  0.4386,  0.5655,  1.2266,  0.7068]])

Tensor a[1:, 2:4]:
tensor([[[ 0.4526, -1.0238, -1.7574,  0.4045,  0.0

a[1] có shape là [4, 5]. a[1:, 2:4] có shape là [2, 2, 5].

In [9]:
print(a[:, :, -2:])

tensor([[[ 1.5262,  0.5968],
         [-2.9484, -0.5147],
         [-0.4894, -1.2085],
         [ 0.9853, -0.5922]],

        [[ 0.3552,  0.7421],
         [-0.7558, -0.2232],
         [ 0.4045,  0.0904],
         [ 1.2266,  0.7068]],

        [[-1.4998, -1.0209],
         [ 1.3314, -0.8596],
         [-1.4476, -0.9064],
         [-0.2195,  1.2015]]])


BÀI 5

In [10]:
import torch

a = torch.randn(3, 4)

print("Original:")
print(a)

print("\nFlatten:")
print(a.ravel())

print("\nReshape 3 x 2 x 2:")
print(a.reshape(3, 2, 2))

Original:
tensor([[ 0.4634, -1.2345, -0.1858,  1.2207],
        [-0.9845, -0.2213, -0.2125,  1.0299],
        [-1.4181, -0.1899,  0.0092, -0.6725]])

Flatten:
tensor([ 0.4634, -1.2345, -0.1858,  1.2207, -0.9845, -0.2213, -0.2125,  1.0299,
        -1.4181, -0.1899,  0.0092, -0.6725])

Reshape 3 x 2 x 2:
tensor([[[ 0.4634, -1.2345],
         [-0.1858,  1.2207]],

        [[-0.9845, -0.2213],
         [-0.2125,  1.0299]],

        [[-1.4181, -0.1899],
         [ 0.0092, -0.6725]]])


In [5]:
a = torch.arange(12).reshape(3, 4)

print("Tensor ban đầu:")
print(a)

b = a.reshape(2, 6)

print("\nTensor sau khi reshape thành 2 × 6:")
print(b)

Tensor ban đầu:
tensor([[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11]])

Tensor sau khi reshape thành 2 × 6:
tensor([[ 0,  1,  2,  3,  4,  5],
        [ 6,  7,  8,  9, 10, 11]])


BÀI 6

In [6]:
import torch
a = torch.randn(2, 3)
b = torch.randn(2, 3)
print("a:")
print(a)
print("b:")
print(b)
print("\na + b:")
print(a + b)
print("\na / b:")
print(a / b)
print("\na^2:")
print(a ** 2)
print("\nMatrix multiplication:")
print(a @ b.T)

a:
tensor([[ 0.0627,  0.0696,  1.2341],
        [ 1.5498, -1.1412,  0.2026]])
b:
tensor([[-0.6394, -1.9830,  0.0537],
        [-0.5455,  0.6164,  1.1726]])

a + b:
tensor([[-0.5767, -1.9134,  1.2878],
        [ 1.0043, -0.5248,  1.3752]])

a / b:
tensor([[-0.0980, -0.0351, 22.9842],
        [-2.8411, -1.8513,  0.1728]])

a^2:
tensor([[0.0039, 0.0048, 1.5230],
        [2.4019, 1.3023, 0.0410]])

Matrix multiplication:
tensor([[-0.1118,  1.4558],
        [ 1.2829, -1.3113]])


a*b là nhân từng phần tử
a @ b.T là nhân ma trận

Đáp án:
shape(a @ b.T ) = torch.Size([2,2])

BÀI 7

In [7]:
import torch

a = torch.randn(3, 4)

print(a)

print("\nMean:")
print(torch.mean(a, dim=0))

print("\nStandard deviation:")
print(torch.std(a, dim=0))

print("\nCumulative sum:")
print(torch.cumsum(a, dim=0))

tensor([[-0.6658,  1.8081,  0.6689, -0.6843],
        [-0.1623,  0.9623,  0.3195, -1.7740],
        [ 1.3878, -0.7692,  0.1769, -0.7003]])

Mean:
tensor([ 0.1865,  0.6671,  0.3884, -1.0529])

Standard deviation:
tensor([1.0703, 1.3138, 0.2531, 0.6246])

Cumulative sum:
tensor([[-0.6658,  1.8081,  0.6689, -0.6843],
        [-0.8282,  2.7704,  0.9884, -2.4584],
        [ 0.5596,  2.0012,  1.1653, -3.1587]])


In [8]:
import torch

a = torch.randn(3, 4)

print(a)

print("\nMean:")
print(torch.mean(a, dim=0))

print("\nStandard deviation:")
print(torch.std(a, dim=0))

print("\nCumulative sum:")
print(torch.cumsum(a, dim=1))

tensor([[-0.1304,  0.4151,  0.7047, -1.4809],
        [ 0.5310, -0.7987,  0.3372,  0.4969],
        [-0.4077,  0.8134, -0.2227, -0.2286]])

Mean:
tensor([-0.0024,  0.1433,  0.2731, -0.4042])

Standard deviation:
tensor([0.4823, 0.8397, 0.4670, 1.0005])

Cumulative sum:
tensor([[-0.1304,  0.2847,  0.9894, -0.4915],
        [ 0.5310, -0.2677,  0.0695,  0.5664],
        [-0.4077,  0.4057,  0.1829, -0.0457]])


dim=0 → tính theo chiều dọc/cột.

dim=1 → tính theo chiều ngang/hàng.

BÀI 8

In [9]:
import torch

x = torch.tensor(3.6, requires_grad=True)

y = x * x

y.backward()

print("x =", x)
print("y =", y)
print("x.grad =", x.grad)

x = tensor(3.6000, requires_grad=True)
y = tensor(12.9600, grad_fn=<MulBackward0>)
x.grad = tensor(7.2000)


Ta có:
y=x^2 
Đạo hàm:
dy/dx = 2x
Tại
x=3.6
Thì 
dy/dx = 2*3.6 = 7.2
Vì Vậy:
x.grad = tensor(7.2000)

In [10]:
import torch

x = torch.tensor(3.6, requires_grad=True)

y = 3 * x**2 + 2 * x + 1

y.backward()

print("y =", y)
print("dy/dx =", x.grad)

y = tensor(47.0800, grad_fn=<AddBackward0>)
dy/dx = tensor(23.6000)


BÀI 9

In [12]:
import numpy as np
import torch

polynomial = np.poly1d([1, 2, 3])

N = 20
X = np.random.randn(N, 1) * 5
Y = polynomial(X)

XX = np.hstack([
    X * X,
    X,
    np.ones_like(X)
])

w = torch.randn(3, 1, requires_grad=True)

x = torch.tensor(XX, dtype=torch.float32)
y = torch.tensor(Y, dtype=torch.float32)

optimizer = torch.optim.NAdam([w], lr=0.01)

print("Initial coefficients:")
print(w)

for epoch in range(1000):
    optimizer.zero_grad()
    y_pred = x @ w
    mse = torch.mean(torch.square(y - y_pred))
    mse.backward()
    optimizer.step()

print("Learned coefficients:")
print(w)

Initial coefficients:
tensor([[ 0.1064],
        [-0.2239],
        [ 0.5871]], requires_grad=True)
Learned coefficients:
tensor([[1.0070],
        [1.9939],
        [2.2446]], requires_grad=True)


1. So sánh ba hệ số học được với [1, 2, 3]:

Ba hệ số học được lần lượt là 1.0070, 1.9939 và 2.2446. Hai hệ số đầu gần với giá trị thực là 1 và 2, còn hệ số thứ ba chưa đạt gần 3.

2. Giải thích:

optimizer.zero_grad(): xóa gradient của lần lặp trước.
mse.backward(): tính gradient của hàm mất mát bằng Autograd.
optimizer.step(): cập nhật các hệ số w dựa trên gradient.

BÀI 10

In [13]:
import random
import torch

A = torch.tensor(random.random(), requires_grad=True)
B = torch.tensor(random.random(), requires_grad=True)
C = torch.tensor(random.random(), requires_grad=True)
D = torch.tensor(random.random(), requires_grad=True)

EPOCHS = 2000

optimizer = torch.optim.NAdam([A, B, C, D], lr=0.01)

for _ in range(EPOCHS):
    y1 = A + B - 9
    y2 = C - D - 1
    y3 = A + C - 8
    y4 = B - D - 2

    sqerr = y1*y1 + y2*y2 + y3*y3 + y4*y4

    optimizer.zero_grad()
    sqerr.backward()
    optimizer.step()

print("A =", A)
print("B =", B)
print("C =", C)
print("D =", D)

A = tensor(4.8166, requires_grad=True)
B = tensor(4.1833, requires_grad=True)
C = tensor(3.1834, requires_grad=True)
D = tensor(2.1834, requires_grad=True)


Kết quả: A = 4.8166, B = 4.1833, C = 3.1834, D = 2.1834. Khi thay vào bốn phương trình, các phương trình đều thỏa mãn (sai số rất nhỏ do tính toán số thực).

Nhận xét: Bài toán có nhiều nghiệm vì hệ phương trình không xác định duy nhất cả bốn ẩn, do đó có thể tồn tại nhiều bộ giá trị A, B, C, D cùng thỏa mãn hệ phương trình.

BÀI TẬP VẬN DỤNG
BÀI 11


In [14]:
import torch

X = torch.arange(24)

print("X ban đầu:")
print(X)

print("\n4 x 6:")
print(X.reshape(4, 6))

print("\n2 x 12:")
print(X.reshape(2, 12))

print("\n2 x 3 x 4:")
print(X.reshape(2, 3, 4))

X ban đầu:
tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
        18, 19, 20, 21, 22, 23])

4 x 6:
tensor([[ 0,  1,  2,  3,  4,  5],
        [ 6,  7,  8,  9, 10, 11],
        [12, 13, 14, 15, 16, 17],
        [18, 19, 20, 21, 22, 23]])

2 x 12:
tensor([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11],
        [12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23]])

2 x 3 x 4:
tensor([[[ 0,  1,  2,  3],
         [ 4,  5,  6,  7],
         [ 8,  9, 10, 11]],

        [[12, 13, 14, 15],
         [16, 17, 18, 19],
         [20, 21, 22, 23]]])


In [15]:
A = torch.randn(3, 4)
B = torch.randn(3, 4)

print("\nA + B:")
print(A + B)

print("\nA * B:")
print(A * B)

print("\nA / B:")
print(A / B)

print("\nTrung bình theo cột:")
print(torch.mean(A, dim=0))

print("\nĐộ lệch chuẩn theo cột:")
print(torch.std(A, dim=0))


A + B:
tensor([[-1.1040,  0.9696,  1.4552, -0.1935],
        [-0.2294, -0.5786,  1.5118,  0.2950],
        [ 1.7477, -0.2817, -2.0673, -3.5835]])

A * B:
tensor([[-1.1458, -0.6052, -0.1543, -0.0174],
        [-1.4594,  0.0327,  0.4665,  0.0214],
        [ 0.7609, -0.1045,  0.9448,  3.1492]])

A / B:
tensor([[ -0.3714,  -0.3081, -15.6569,  -0.2570],
        [ -0.8273,   0.1231,   2.4987,   0.7781],
        [  0.8873,  -0.4291,   0.4923,   1.3200]])

Trung bình theo cột:
tensor([ 0.8576, -0.0945,  0.6507, -0.6143])

Độ lệch chuẩn theo cột:
tensor([0.2254, 0.3229, 1.1783, 1.2341])


In [16]:
x = torch.tensor(2.0, requires_grad=True)

y = 2*x**2 + 5*x + 3

y.backward()

print("\ny =", y)
print("dy/dx =", x.grad)


y = tensor(21., grad_fn=<AddBackward0>)
dy/dx = tensor(13.)


Autograd tính được \(dy/dx=13\), trùng với kết quả tính thủ công.

In [17]:
import numpy as np
import torch

polynomial = np.poly1d([2, -3, 5])

N = 20
X = np.random.randn(N, 1) * 5
Y = polynomial(X)

XX = np.hstack([
    X * X,
    X,
    np.ones_like(X)
])

w = torch.randn(3, 1, requires_grad=True)

x = torch.tensor(XX, dtype=torch.float32)
y = torch.tensor(Y, dtype=torch.float32)

optimizer = torch.optim.NAdam([w], lr=0.01)

for epoch in range(1000):
    optimizer.zero_grad()

    y_pred = x @ w

    mse = torch.mean(torch.square(y - y_pred))

    mse.backward()

    optimizer.step()

print("Learned coefficients:")
print(w)

Learned coefficients:
tensor([[ 2.1018],
        [-3.0348],
        [ 0.4584]], requires_grad=True)


Ba hệ số học được lần lượt gần 2, -3, 5, đúng với các hệ số của đa thức \(y=2x^2-3x+5\).